File: examples/rust/matrices/src/matrix.rs
use std::thread;

#[derive(Debug, Clone)]
pub struct Matrix(pub Vec<Vec<f64>>);

impl Matrix {
    pub fn rows(&self) -> usize { self.0.len() }
    pub fn columns(&self) -> usize { self.0[0].len() }

    pub fn sum_serial(&self) -> f64 {
        let mut result = 0.0;
        for row in &self.0 {
            result += self.add_row(row);
        }
        return result;
    }

    pub fn sum_parallel(&self) -> f64 {
       thread::scope(|s| {
            let mut threads = Vec::new();
            for row in &self.0 {
                threads.push(s.spawn(move || self.add_row(&row)));
            }

            let mut result = 0.0;
            for thread in threads {
                result += thread.join().unwrap();
            }
            result
        })
    }

    fn add_row(&self, row: &Vec<f64>) -> f64 {
        let mut result = 0.0;
        for v in row {
            result += v;
        }
        return result
    }

    pub fn add_serial(&self, other: &Matrix) -> Matrix {
        let rows = self.rows();
        let cols = self.columns();
        let mut result = Vec::new();

        for i in 0..rows {
            let mut row = Vec::new();
            for j in 0..cols {
                row.push(self.0[i][j] + other.0[i][j]);
            }
            result.push(row);
        }
        Matrix(result)
    }

    pub fn add_parallel(&self, other: &Matrix) -> Matrix {
        let rows = self.rows();
        let cols = self.columns();

        thread::scope(|s| {
            let threads: Vec<_> = (0..rows)
                .map(|i| {
                    s.spawn(move || {
                        (0..cols).map(|j| self.0[i][j] + other.0[i][j]).collect()
                    })
                })
                .collect();

            Matrix(threads.into_iter()
                .map(|t| t.join().unwrap())
                .collect()
            )
        })
    }


}

File: examples/java/matrices/src/Matrix.java
import java.util.ArrayList;
import java.util.function.Supplier;

public class Matrix {
    private final double[][] values;

    public Matrix(double[][] values) {
        this.values = values;
    }

    public double sum() {
        var result = 0.0;
        for (double[] value : values) {
            result += addRow(value);
        }
        return result;
    }

    public double sumParallel() {
        var threads = new ArrayList<ThreadValue<Double>>();
        for (double[] value : values) {
            ThreadValue<Double> t = new ThreadValue<>(() -> addRow(value));
            t.start();
            threads.add(t);
        }
        // ---- Aca calcula
        for (var thread : threads) {
            try {
                thread.join();
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }
        double result = 0.0;
        for (var t : threads)
            result += t.getValue();

        return result;

    }

    public Matrix addSerial(Matrix other) {
        double[][] result = new double[values.length][];
        for (int i = 0; i < values.length; i++) {
            int cols = values[i].length;
            var row = new double[cols];
            for (int j = 0; j < cols; j++) {
                row[j] = values[i][j] + other.values[i][j];
            }
            result[i] = row;
        }
        return new Matrix(result);
    }
    public Matrix addParallel(Matrix other) {
        throw new RuntimeException("Implement me!");
    }

    private double addRow(double[] value) {
        var result = 0.0;
        for (double v : value) {
            result += v;
        }
        return result;
    }

    static class ThreadValue<T> extends Thread {
        private final Supplier<T> expression;
        private T result;

        public ThreadValue(Supplier<T> expression) {
            this.expression = expression;
        }
        public void run() { result = expression.get(); }
        public T getValue() { return result; }
    }


}